consensus: Store transaction nVersion as uint32_t

achow101 opened the pull request:
Given that the use of a transaction's nVersion is always as an unsigned int, it doesn't make sense to store it as signed and then cast it to unsigned everywhere it is used and displayed.

Since a few alternative implementations have recently been revealed to have made an error with this signedness that would have resulted in consensus failure, I think it makes sense for us to just make this always unsigned to make it clear that the version is treated as unsigned. This would also help us avoid future potential issues with signedness of this value.

I believe that this is safe and does not actually change what transactions would or would not be considered both standard and consensus valid. Within consensus, the only use of the version in consensus is in BIP68 validation which was already casting it to uint32_t. Within policy, although it is used as a signed int for the transaction version number check, I do not think that this change would change standardness. Standard transactions are limited to the range [1, 2]. Negative numbers would have fallen under the < 1 condition, but by making it unsigned, they are still non-standard under the > 2 condition.

Unsigned and signed ints are serialized and unserialized the same way so there is no change in serialization.

achow101 committed:
consensus: Store transaction nVersion as uint32_t

Given that the use of a transaction's nVersion is always as an unsigned
int, it doesn't make sense to store it as signed and then cast it to
unsigned.

maflcko reviewed (approved):
Agree that a rename of the field should be considered, but my preference would be just `version`, as opposed to `unsigned_version`. Otherwise, the code might be annoyingly verbose, even when it is clear to everyone that the transaction version is unsigned. (Maybe in a separate commit, so that if the changes are touching too many lines, or reviewers don't like it, it can be dropped again)

vostrnad commented:
Concept ACK

As someone quite familiar with Bitcoin's consensus rules but largely unfamiliar with Bitcoin Core's internals, I had no idea until recently that the version was stored as a signed integer. This should bring it in line with expectations of most future contributors.

Also negative versions are just weird.

vostrnad commented on src/consensus/tx_verify.cpp:
```diff
@@ -51,8 +51,7 @@ std::pair<int, int64_t> CalculateSequenceLocks(const CTransaction &tx, int flags
     // wouldn't support BIP 68.
-    bool fEnforceBIP68 = static_cast<uint32_t>(tx.nVersion) >= 2
-                      && flags & LOCKTIME_VERIFY_SEQUENCE;
+    bool fEnforceBIP68 = tx.nVersion >= 2 && flags & LOCKTIME_VERIFY_SEQUENCE;
```
Comment above needs updating.

achow101 replied:
Removed it.

fanquake commented:
> So enabling that compiler warning may not be possible without changing a lot more code.

Yea. See here for example output compiling master with GCC 13.2.0 + -Wsign-conversion: https://gist.github.com/fanquake/8e7a49dc1968afd07d89e822ffb4adac.

shaavan commented on test/functional/rpc_rawtransaction.py:
```diff
@@ -478,6 +478,20 @@ def transaction_version_number_tests(self):
         decrawtx = self.nodes[0].decoderawtransaction(rawtx)
         assert_equal(decrawtx['version'], 0x7fffffff)
 
```
Question: If we are removing the "signedness" of the tx.nVersion, should we think about removing the test cases that made sense for the signed nVersion?

achow101 replied:
I think it makes sense to leave them in to make sure that everything still matches previous versions.

shaavan reviewed (commented):
Concept ACK

<details>
<summary>
Notes:
</summary>

1. Change the type of nVersion for transactions from `int32_t` → `uint32_t`
2. Appropriately remove casts, from the codebase
3. Introduce test cases to properly address the change in behavior
4. Appropriately change the format form `<i` to `<I` indicating the change from signed little-endian to unsigned little-endian when serializing and deserializing tx.nVersion. 
</details>

Suggestions:

1. In `void static RandomTransaction(CMutableTransaction& tx, bool fSingle)` in sighash_tests.cpp file:
    
    `tx.nVersion = int(InsecureRand32());` → `tx.nVersion = InsecureRand32();` since the return type of the function is `uint32_t`

achow101 committed:
refactor: Rename CTransaction::nVersion to version

In order to ensure that the change of nVersion to a uint32_t in the
previous commit has no effect, rename nVersion to version in this commit
so that reviewers can easily spot if a spot was missed or if there is a
check somewhere whose semantics have changed.

achow101 commented on src/core_write.cpp:
```diff
@@ -176,7 +176,7 @@ void TxToUniv(const CTransaction& tx, const uint256& block_hash, UniValue& entry
     // Transaction version is actually unsigned in consensus checks, just signed in memory,
     // so cast to unsigned before giving it to the user.
-    entry.pushKV("version", static_cast<int64_t>(static_cast<uint32_t>(tx.nVersion)));
+    entry.pushKV("version", static_cast<int64_t>(tx.nVersion));
```
Done, removed the comment.

achow101 commented on test/functional/feature_taproot.py:
```diff
@@ -1410,7 +1410,7 @@ def test_spenders(self, node, spenders, input_counts):
             # Construct CTransaction with random nVersion, nLocktime
             tx = CTransaction()
-            tx.nVersion = random.choice([1, 2, random.randint(-0x80000000, 0x7fffffff)])
+            tx.nVersion = random.choice([1, 2, random.randrange(0xffffffff)])
```
Done

achow101 commented:
I've added a commit that renames `nVersion` to just `version`. Note that it is not a scripted diff since a bunch of different variables share the same name.

dergoegge commented:
* [CURRENT_VERSION](https://github.com/bitcoin/bitcoin/blob/3a41d8ca8481ce7f85b1c9a9638d63e8e5fa5285/src/primitives/transaction.h#L299) should probably also change to `uint32_t`

*  CI failes [here](https://github.com/bitcoin/bitcoin/blob/3a41d8ca8481ce7f85b1c9a9638d63e8e5fa5285/src/test/fuzz/util.cpp#L46-L48) due to UB, can be fixed by changing to `ConsumeIntegral<uint32_t>()`

achow101 commented on src/bitcoin-tx.cpp:
```diff
@@ -210,7 +210,7 @@ static void MutateTxVersion(CMutableTransaction& tx, const std::string& cmdVal)
     }
 
-    tx.nVersion = (int) newVersion;
+    tx.version = (int) newVersion;
```
Since `newVersion` is an `int64_t`, I think this still needs to be cast to `uint32_t`.

achow101 commented on src/primitives/transaction.cpp:
```diff
@@ -63,8 +63,8 @@ std::string CTxOut::ToString() const
 
-CMutableTransaction::CMutableTransaction() : nVersion(CTransaction::CURRENT_VERSION), nLockTime(0) {}
-CMutableTransaction::CMutableTransaction(const CTransaction& tx) : vin(tx.vin), vout(tx.vout), nVersion(tx.nVersion), nLockTime(tx.nLockTime) {}
+CMutableTransaction::CMutableTransaction() : version(CTransaction::CURRENT_VERSION), nLockTime(0) {}
```
Done

achow101 commented on src/primitives/transaction.cpp:
```diff
@@ -63,8 +63,8 @@ std::string CTxOut::ToString() const
-CMutableTransaction::CMutableTransaction() : nVersion(CTransaction::CURRENT_VERSION), nLockTime(0) {}
-CMutableTransaction::CMutableTransaction(const CTransaction& tx) : vin(tx.vin), vout(tx.vout), nVersion(tx.nVersion), nLockTime(tx.nLockTime) {}
+CMutableTransaction::CMutableTransaction() : version(CTransaction::CURRENT_VERSION), nLockTime(0) {}
+CMutableTransaction::CMutableTransaction(const CTransaction& tx) : vin(tx.vin), vout(tx.vout), version(tx.version), nLockTime(tx.nLockTime) {}
```
Done

achow101 commented:
> * [CURRENT_VERSION](https://github.com/bitcoin/bitcoin/blob/3a41d8ca8481ce7f85b1c9a9638d63e8e5fa5285/src/primitives/transaction.h#L299) should probably also change to `uint32_t`
> 
>     * CI failes [here](https://github.com/bitcoin/bitcoin/blob/3a41d8ca8481ce7f85b1c9a9638d63e8e5fa5285/src/test/fuzz/util.cpp#L46-L48) due to UB, can be fixed by changing to `ConsumeIntegral<uint32_t>()`

Updated these.

naumenkogs commented:
Concept ACK

achow101 commented on src/bitcoin-tx.cpp:
```diff
@@ -210,7 +210,7 @@ static void MutateTxVersion(CMutableTransaction& tx, const std::string& cmdVal)
     }
 
-    tx.nVersion = (int) newVersion;
+    tx.nVersion = static_cast<uint32_t>(newVersion);
```
Ah, didn't realize we had that. Done

achow101 commented on src/primitives/transaction.h:
```diff
@@ -296,7 +296,7 @@ class CTransaction
 public:
     // Default transaction version.
-    static const int32_t CURRENT_VERSION=2;
+    static const uint32_t CURRENT_VERSION=2;
```
Done

maflcko reviewed (approved):
ACK 6acfcbfe2487f683e8f62606d195a9974bc2234f 🌵

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: ACK 6acfcbfe2487f683e8f62606d195a9974bc2234f 🌵
rWI9ANXeLccWSbPTbyKsnso1o6VZc+jr+YkpUEQVUs4ZzPer/ucIHhkSP4LZEz8sZYOUbXyFmxFoSAGyuvN4DA==
```

</details>

naumenkogs commented on src/test/transaction_tests.cpp:
```diff
@@ -780,7 +780,7 @@ BOOST_AUTO_TEST_CASE(test_IsStandard)
 
     // Disallowed nVersion
-    t.nVersion = -1;
+    t.nVersion = static_cast<uint32_t>(-1);
```
ca7090d43fab2013bd396d683f4cf03062666b14

Does this still look meaningful? 
If you want to test a large number, then better check against `TX_MAX_STANDARD_VERSION + 1`?

achow101 replied:
Changed to `std::numeric_limits<uint32_t>::max()`

naumenkogs reviewed (approved):
ACK 6acfcbfe2487f683e8f62606d195a9974bc2234f

achow101 commented on test/functional/test_framework/messages.py:
```diff
@@ -581,7 +581,7 @@ def __init__(self, tx=None):
 
     def deserialize(self, f):
-        self.nVersion = int.from_bytes(f.read(4), "little", signed=True)
+        self.nVersion = int.from_bytes(f.read(4), "little", signed=False)
```
Done as suggested.

achow101 commented on test/functional/test_framework/script.py:
```diff
@@ -739,7 +739,7 @@ def SegwitV0SignatureMsg(script, txTo, inIdx, hashtype, amount):
 
     ss = bytes()
-    ss += struct.pack("<i", txTo.nVersion)
+    ss += struct.pack("<I", txTo.nVersion)
```
Done as suggested.

achow101 commented on src/primitives/transaction.cpp:
```diff
@@ -117,7 +117,7 @@ std::string CTransaction::ToString() const
     std::string str;
     str += strprintf("CTransaction(hash=%s, ver=%d, vin.size=%u, vout.size=%u, nLockTime=%u)\n",
```
Done

achow101 commented on test/functional/p2p_segwit.py:
```diff
@@ -1164,7 +1164,7 @@ def serialize_with_witness(self):
                     flags |= 1
                 r = b""
-                r += struct.pack("<i", self.nVersion)
+                r += struct.pack("<i", self.version)
```
Done

maflcko reviewed (approved):
Left some nits. Lgtm

re-ACK 4542a4d76e7598ac67fa4f154b13b2fa32050dd2  🌷

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: re-ACK 4542a4d76e7598ac67fa4f154b13b2fa32050dd2  🌷
gZ56+jV2aJ8Pwn0w4jdGsrhbCyYQh5Syoq40v5jlZBUMztqTt5uUIu0rYOSOzJFKOchO+zQALuivCEUPn5W5AA==
```

</details>

achow101 commented on test/functional/feature_bip68_sequence.py:
```diff
@@ -410,6 +410,5 @@ def test_version2_relay(self):
         mini_wallet = MiniWallet(self.nodes[1])
         mini_wallet.send_self_transfer(from_node=self.nodes[1], version=2)
 
-
```
Fixed

achow101 commented:
Changed all of the remaining tx.nVersion that I could find.

naumenkogs commented:
ACK 4a45b28a3f258bf156eca980aace69e0a90554cd

vostrnad commented:
Since DrahtBot is incessantly pinging me for review, here it is:

ACK 4a45b28a3f258bf156eca980aace69e0a90554cd

I reviewed the code with my somewhat limited knowledge of C++ and didn't find any problems. I also checked that all instances of `nVersion` that referred to the transaction version have been renamed, except for the command line argument in bitcoin-tx where it would be a breaking change.

shaavan reviewed (approved):
ACK 4a45b28a3f258bf156eca980aace69e0a90554cd

As far as the internal tx nVersion variables go, all the instances of it are rightly converted to uint32_t. All the typecasting is appropriately modified (or removed) for clarity. And we moved to using `int.to_bytes` instead of `struct.pack` in the test which allows for a cleaner, more readable code by using the function more suitable for integer conversion.

The only question I have is I notice that though the instances of internal tx `nVersions` are modified, the command line call of it still uses `nVersion` as the variable:

https://github.com/bitcoin/bitcoin/blob/1ac627c485a43e50a9a49baddce186ee3ad4daad/src/bitcoin-tx.cpp#L57

https://github.com/bitcoin/bitcoin/blob/1ac627c485a43e50a9a49baddce186ee3ad4daad/src/bitcoin-tx.cpp#L713-L714

This makes sense, as we don’t want to break compatibility with code that already utilizes this variable in their code and scripts. What I am curious about is whether we will ever be moving the using `version` as the name of the command line variable as well.

P.S.: On a side, I must say that DrahtBot is a persistent boi! :)

achow101 commented:
Rebased

vostrnad commented:
re-ACK cc0d6bd73e67abde657cb8fb38c0e18de316d337

achow101 commented:
@maflcko @vostrnad @shaavan @naumenkogs reACKs would be appreciated

shaavan reviewed (approved):
ACK 2431ecff2fb6f4b68d3de0ea36e3bcc4403e94ba 🚀

shaavan reviewed (commented):
ACK 659663af32f02c570e334e8f375fd5f5737258d7 ⚡️

glozow commented:
concept ACK + lgtm 659663af32

Seems like this is a remaining mention of `CTransaction::nVersion`?
https://github.com/bitcoin/bitcoin/blob/659663af32f02c570e334e8f375fd5f5737258d7/src/compressor.h#L57-L60

Also noting that #29496 unfortunately conflicts and I would advocate for merging that first... but I'm happy to re-review

glozow commented:
ACK 429ec1aaaa

Thank you for rebasing!

shaavan reviewed (commented):
ACK 429ec1aaaaafab150f11e27fcf132a99b57c4fc7 💯
